### PR TITLE
[Parser] Support "<" unary operator in #if swift() and #if compiler() expressions

### PR DESCRIPTION
**Pull request by danielmartin**
**Swift-evolution proposal:** https://github.com/apple/swift-evolution/pull/877
Resolves #49401.

Until now, only ">=" was supported in #if swift() expressions, for example:

```
#if swift(>=2.1)
#endif
```

This means that if we want to evaluate code only when the language version is
less than a particular version we need to do the following:

```
#if !swift(>=2.1)
#endif
```

An alternative to make this more readable (the "!" can be easily missed in a code
review) is to introduce another supported unary operator, "<". The previous
example could be rewritten like this:

```
#if swift(<2.1)
#endif
```

This commit adds support for that unary operator, along with some tests.

### CODE REVIEW

**Comment by CodaFi** [comment]
Why not just lift the restriction on operators altogether.

**Comment by jrose-apple** [comment]
We don't want to support `>` or `<=` because they don't account for patch versions. That is, if a problem is not fixed in a patch, you'd need to update your condition, and if it *is* fixed in a patch, you can use `>=` or `<` instead.

File: lib/Parse/ParseIfConfig.cpp [hunk 1]
```diff
@@ -53,6 +53,17 @@ static StringRef extractExprSource(SourceManager &SM, Expr *E) {
   return SM.extractText(Range);
 }
 
+static bool isValidPrefixUnaryOperator(llvm::Optional<StringRef> &UnaryOperator) {
+  return UnaryOperator != None && (UnaryOperator.getValue() == ">=" || UnaryOperator.getValue() == "<");
+}
+
+static bool isValidVersion(version::Version &Version, version::Version &ExpectedVersion, StringRef UnaryOperator) {
```
Comment (jrose-apple) [inline lib/Parse/ParseIfConfig.cpp]: Nitpick: These should be `const` references, and please wrap at 80 columns. (`clang-format` should be set up correctly for this repo.)

File: lib/Parse/ParseIfConfig.cpp [hunk 2]
```diff
@@ -53,6 +53,17 @@ static StringRef extractExprSource(SourceManager &SM, Expr *E) {
   return SM.extractText(Range);
 }
 
+static bool isValidPrefixUnaryOperator(llvm::Optional<StringRef> &UnaryOperator) {
```
Comment (jrose-apple) [inline lib/Parse/ParseIfConfig.cpp]: Nitpick: We import `Optional` into the `swift` namespace like Clang does, so you can just use `Optional<StringRef>` here. Probably no need to pass by reference either.

**Review by jrose-apple — COMMENTED** [review]
Actual implementation looks pretty good, just a few nitpicks. Merging needs to wait for the review, of course, or for the Core Team to formally say this doesn't need one (still possible given Chris's positive feedback on the forum thread).

File: lib/Parse/ParseIfConfig.cpp (hunk 1 above)
Comment (danielmartin) [reply lib/Parse/ParseIfConfig.cpp]: Ah, thanks. I've run `git clang-format` to not pollute the diff.

File: lib/Parse/ParseIfConfig.cpp [hunk 3]
```diff
@@ -448,13 +462,16 @@ class EvaluateIfConfigCondition :
       return thisVersion >= Val;
     } else if ((KindName == "swift") || (KindName == "compiler")) {
       auto PUE = cast<PrefixUnaryExpr>(Arg);
+      auto UDRE = getDeclRefStr(PUE->getFn());
```
Comment (rintaro) [inline lib/Parse/ParseIfConfig.cpp]: Nitpick: `UDRE` (UnresolvedDeclRefExpr) is a little odd here. I suggest `PrefixName` as per `ValidateIfConfigCondition:: visitCallExpr`.

File: test/Parse/ConditionalCompilation/language_version.swift [hunk 4]
```diff
@@ -34,13 +48,21 @@
   %#^*&
 #endif
 
-#if swift(">=7.1") // expected-error {{unexpected platform condition argument: expected a unary comparison, such as '>=2.2'}}
+#if !swift(<1000.0)
```
Comment (tkremenek) [inline test/Parse/ConditionalCompilation/language_version.swift]: Please also add test cases for `#if compiler`, whose behavior is also changed by this PR.

**Review by tkremenek — CHANGES_REQUESTED** [review]

**Comment by tkremenek** [comment]
SE-0224 has been approved, so this is clear to land once this PR is approved.

File: test/Parse/ConditionalCompilation/language_version.swift (hunk 4 above)
Comment (danielmartin) [reply test/Parse/ConditionalCompilation/language_version.swift]: I noticed that tests for `#if compiler` were missing in the codebase, so I added tests for both operators in a new test file (https://github.com/apple/swift/pull/17960/commits/ccbc3f7dcc8bb41f3ab588b3167f1052875b4095)

File: test/Parse/ConditionalCompilation/compiler.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// RUN: %target-typecheck-verify-swift
```
Comment (benrimmington) [inline test/Parse/ConditionalCompilation/compiler.swift]: Could this *also* be run in Swift 4 mode, to check that `#if compiler` is not affected?

```diff
 // RUN: %target-typecheck-verify-swift
+// RUN: %target-typecheck-verify-swift -swift-version 4
```

**Comment by jrose-apple** [comment]
@tkremenek, I think this is waiting on you.

@swift-ci Please test

**Comment by tkremenek** [comment]
Re-running CI against latest master.

@swift-ci test

**Review by tkremenek — APPROVED** [review]

**Comment by benrimmington** [comment]
@tkremenek All checks have passed.

**Pull request merged**